src/screens/game_map.py

```diff
@@ -0,0 +1,681 @@
+import math
+from collections.abc import Callable
+from typing import Any
+
+import pygame
+from pytmx import (
+    TiledMap,
+    TiledTileLayer,
+    TiledObjectGroup,
+    TiledObject,
+    TiledElement
+)
+
+from src.enums import Layer, FarmingTool, InventoryResource
+from src.groups import AllSprites, PersistentSpriteGroup
+from src.gui.interface.emotes import NPCEmoteManager, PlayerEmoteManager
+from src.map_objects import MapObjects
+from src.npc.bases.animal import Animal
+from src.npc.chicken import Chicken
+from src.npc.cow import Cow
+from src.npc.npc import NPC
+from src.npc.setup import AIData
+from src.overlay.soil import SoilLayer
+from src.settings import (
+    SETUP_PATHFINDING, ENABLE_NPCS, TEST_ANIMALS,
+    SCALED_TILE_SIZE, SCALE_FACTOR, TILE_SIZE
+)
+from src.sprites.base import Sprite, AnimatedSprite, CollideableMapObject
+from src.sprites.character import Character
+from src.sprites.drops import DropsManager
+from src.sprites.entities.player import Player
+from src.sprites.objects.tree import Tree
+from src.sprites.setup import ENTITY_ASSETS
+
+
+def _setup_tile_layer(
+        layer: TiledTileLayer,
+        func: Callable[[tuple[int, int], pygame.Surface], None]
+):
+    """
+    Calls func for each tile found in layer
+    :param layer: TiledTileLayer
+    :param func: function(pos, image)
+    """
+    for x, y, image in layer.tiles():
+        x = x * SCALED_TILE_SIZE
+        y = y * SCALED_TILE_SIZE
+        pos = (x, y)
+        func(pos, image)
+
+
+def _setup_object_layer(
+        layer: TiledObjectGroup,
+        func: Callable[[tuple[float, float], TiledObject], Any]
+):
+    """
+    Calls func for each tile found in layer
+    :param layer: TiledTileLayer
+    :param func: function(pos, object) -> object instance
+    :return: All object instances
+    """
+    objects = []
+    for obj in layer:
+        x = obj.x * SCALE_FACTOR
+        y = obj.y * SCALE_FACTOR
+        pos = (x, y)
+        objects.append(func(pos, obj))
+    return objects
+
+
+def _get_element_property(
+        element: TiledElement,
+        property_name: str,
+        callback: Callable[[str], Any],
+        default: Any
+) -> Any:
+    """
+    :param element: Element to retrieve the property value from
+    :param property_name: Name of the property
+    :param callback: Function that should be called with the property value
+                     retrieved from element
+    :param default: Default value to return if callback raises an exception
+    :return: Return value of callback
+    """
+    prop = element.properties.get("layer")
+    if prop:
+        try:
+            return callback(prop)
+        except Exception as e:
+            print(
```

Comment:
In general, it would be better to use `warnings.warn` for this, because just printing may confuse some people (as `warnings.warn` will be printing in `sys.stderr`, so you will be able to split the info more easily

Reply:
Thank you very much! No idea why I've never heard of warnings.warn before 😅